### ISSUE TITLE
🐕 Task: Running the same model in parallel

### ISSUE THREAD

**Issue by miquelduranfrigola** [ersilia-os/ersilia#1223]
### Summary

As we work on running more than one Ersilia model in parallel, @JHlozek highlighted the scenario where we want to run **the same model** in multiple processes/terminals. This would be a very interesting case to consider in repositories like Olinda, where we need to make precalculations across a large set of inputs.

### Objective(s)

1. Run the same model in multiple terminals/processes (i.e. sessions).
2. Optionally, make sure parallelization works both in docker and in conda serving modes.
3. Ideally, parallelization should work both from the CLI (i.e. `ersilia run -i ...`) and from the Python API (`mdl.run(...)`). The Python API parallelization may be more difficult and it is less critical.

### Documentation

No specific documentation available for this, although we should include parallelization as part of our main documentation in Gitbook and the README file.

**Comment by miquelduranfrigola** [comment]
@DhanshreeA what's your take on this? Is this something that may be relatively easy to address?

**Comment by DhanshreeA** [comment]
I am hoping this works out of the box really since all session artifacts go within their respective folders. I'll test this.

**Comment by GemmaTuron** [comment]
Hi @DhanshreeA 

Was this tested? Is this something we could describe better and label as "good first issue"?

**Comment by DhanshreeA** [comment]
@GemmaTuron Right off the bat, this is not a safe good first issue because of several reasons as follows:

1. Presently, even with the implementation of running multiple ersilia models simultaneously across different terminals, we cannot serve the same model across two different terminals because before ersilia starts a model container, it closes all previous containers related to that model. I am not clear on why we built it this way but my understanding is that there is probably safety in doing in this functionality because of the following reason:
- If a user has a container running from an older version of the model's image, stopping that container and re-spawning a new container would ensure the user has access to the latest model code, even if it doesn't always happen in practice because neither do our models get updated very frequently, and we don't always fetch a model before serving it. I think we should rewrite this functionality to stop all containers related to a model only when the model is being fetched, and not when it is being served. The [`PulledDockerImageService`](https://github.com/ersilia-os/ersilia/blob/6b6c6d3f271bf4db7865107320893e9c73b44bb5/ersilia/serve/services.py#L639) class should not call the method [`_stop_all_containers_of_image`](https://github.com/ersilia-os/ersilia/blob/6b6c6d3f271bf4db7865107320893e9c73b44bb5/ersilia/serve/services.py#L720) when the model is being served, which I think can be achieved with a flag. 

2. Parallelization currently does not work within the case of `conda-docker`. It works when all models are either docker based, or conda based, but not in a mixed scenario. This was specifically mentioned as a requirement [here](https://github.com/ersilia-os/ersilia/issues/702#issue-1756694049). There is a potential fix for it [here](https://github.com/ersilia-os/ersilia/issues/1197), but we need to make sure we want this functionality. 

3. I cannot comment on this at all because we need to tackle the Python API completely and make sure it has caught up with the recent developments within Ersilia. I have proposed this as one of the key areas within Outreachy and I will come up with a plan for it.

**Comment by DhanshreeA** [comment]
@miquelduranfrigola we need to revisit this at some point.

**Comment by DhanshreeA** [comment]
This is also something we should revisit next week.

**Comment by GemmaTuron** [comment]
@DhanshreeA this has a deadline for tomorrow. I think there are too many Ersilia Maintanence issues to be completed before tackling new features. Please add a realistic timeline, or leave this on hold

**Comment by miquelduranfrigola** [comment]
At the moment, you can serve the same model from two terminals **but** only one Docker container will stay active. That is, when you serve from the second terminal, the containers served from the first terminal will be destroyed. This does not mean that you cannot run predictions from the first terminal - Ersilia automatically resolves to whatever active container is available for the model of the session.

We may want to reconsider this behaviour in the future but, for now, I believe this is good enough unless we demonstrate that concurrency of API requests to the same container causes problems.
